docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,69 @@
+# Qualifying replicas for read operations
+
+## Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+a replica which is overloaded or dead.
+
+## Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple estimation
+of how responsive a replica is likely to be is implemented as follows:
+ * each time a read request is sent to a replica, a timestamp is recorded
+ * each time a response is received from a replica, a timestamp is recorded
+ * the difference between these two timestamps is a cheap, rough estimate
+   of how long it may take this replica to respond
+Note that this estimation is not precise - it does not determine whether
+a `sent` timestamp and a `responded` timestamp belong to the same request,
+so these values may vary. However, assuming that the cluster is under
+continuous, uniform load, these values would get updated very often, and would
+hopefully approach the approximate round-trip time for a single request.
```

Comment:
Why would they? For a heavily used cluster, we'd be continuously sending requests and receiving responses. A replica could be responding 4 seconds too late, and still have continuous requests and responses.

Reply:
Hm, right. I still haven't run an aws-based workload with a slower node, but that's what I'd most likely discover.